packages/pymedphys_deliverydata/src/pymedphys_deliverydata/utilities/masking.py

```diff
@@ -0,0 +1,124 @@
+# (at your option) any later version (the "AGPL-3.0+").
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+
+import numpy as np
+
+from ..base import _DeliveryDataBase
+
+
+def get_all_masked_delivery_data(delivery_data: _DeliveryDataBase,
+                                 template_gantry_angles, gantry_tol):
+    masks = get_gantry_angle_masks(
+        delivery_data, template_gantry_angles, gantry_tol)
+
+    all_masked_delivery_data = (
+        apply_mask_to_delivery_data(delivery_data, mask)
+        for mask in masks
+    )
+
+    return all_masked_delivery_data
+
+
+def get_gantry_angle_masks(delivery_data: _DeliveryDataBase, gantry_angles,
+                           gantry_tol):
+    masks = [
+        gantry_angle_mask(delivery_data, gantry_angle, gantry_tol)
+        for gantry_angle in gantry_angles
+    ]
+
+    for mask in masks:
+        if np.all(mask == 0):
+            continue
+
+        # TODO: Apply mask by more than just gantry angle to appropriately
+        # extract beam index even when multiple beams have the same gantry
+        # angle
+        assert np.sum(np.abs(np.diff(np.concatenate([
+            [0], mask, [0]])))) == 2, "Duplicate gantry angles not yet supported"
+
+    try:
+        assert np.all(np.sum(masks, axis=0) == 1), (
+            "Not all beams were captured by the gantry tolerance of "
+            " {}".format(gantry_tol)
+        )
+    except AssertionError:
+        print("Allowable gantry angles = {}".format(gantry_angles))
+        gantry = np.array(delivery_data.gantry, copy=False)
+        out_of_tolerance = np.unique(
+            gantry[np.sum(masks, axis=0) == 0]).tolist()
+        print("The gantry angles out of tolerance were {}".format(
+            out_of_tolerance))
+
+        raise
+
+    return masks
+
+
+def gantry_angle_mask(delivery_data, gantry_angle, gantry_angle_tol):
+    near_angle = np.abs(
+        np.array(delivery_data.gantry) - gantry_angle) <= gantry_angle_tol
+    assert np.all(np.diff(np.where(near_angle)[0]) == 1)
+
+    return near_angle
+
+
+def apply_mask_to_delivery_data(delivery_data: _DeliveryDataBase, mask):
```

Comment:
method

Reply:
Also... this function was written before the class was decided to be made.

... if that helps see my reasoning.